h2o-extensions/xgboost/src/main/java/ml/dmlc/xgboost4j/java/XGBoostUpdateTask.java

```diff
@@ -1,193 +1,59 @@
-                      Booster booster,
-                      BoosterParms boosterParms,
-                      int tid,
-                      Map<String, String> workerEnvs) {
-        _nodeToMatrixWrapper = setupTask._nodeToMatrixWrapper;
-        _boosterParms = boosterParms;
-        _boosterHolder = wrap(booster);
+    public XGBoostUpdateTask(XGBoostSetupTask setupTask, int tid) {
+        super(setupTask);
         _tid = tid;
-        rabitEnv.putAll(workerEnvs);
     }
 
     @Override
-    protected void setupLocal() {
-        if(H2O.ARGS.client) {
-            return;
-        }
-
-        // We need to verify that the xgboost is available on the remote node
-        if (!ExtensionManager.getInstance().isCoreExtensionEnabled(XGBoostExtension.NAME)) {
-            throw new IllegalStateException("XGBoost is not available on the node " + H2O.SELF);
-        }
-        try {
-            PersistentDMatrix.Wrapper wrapper = (PersistentDMatrix.Wrapper) _nodeToMatrixWrapper.get(H2O.SELF.toString());
-            if (wrapper == null)
-                return;
-            DMatrix matrix = wrapper.get();
-            update(matrix);
-        } catch (XGBoostError xgBoostError) {
-            try {
-                Rabit.shutdown();
-            } catch (XGBoostError xgBoostError1) {
-                xgBoostError1.printStackTrace();
-            }
-            xgBoostError.printStackTrace();
-            throw new IllegalStateException("Failed XGBoost training.", xgBoostError);
-        }
+    protected void execute() {
+        Booster booster = XGBoostUpdater.doUpdate(_modelKey, _tid);
+        if (booster == null)
+            throw new IllegalStateException("Boosting iteration didn't produce a valid Booster.");
     }
 
-    private void update(DMatrix trainMat) throws XGBoostError {
-        rabitEnv.put("DMLC_TASK_ID", String.valueOf(H2O.SELF.index()));
-
-        try {
-            // DON'T put this before createParams, createPrams calls train() which isn't supposed to be distributed
-            // just to check if we have GPU on the machine
-            Rabit.init(rabitEnv);
-
-            final Booster booster;
-            if (! _boosterHolder.hasBooster()) {
-                HashMap<String, DMatrix> watches = new HashMap<>();
-                booster = ml.dmlc.xgboost4j.java.XGBoost.train(trainMat,
-                        _boosterParms.get(),
-                        0,
-                        watches,
-                        null,
-                        null);
-            } else {
-                // Do one iteration
-                booster = _boosterHolder.get(_boosterParms);
-                booster.update(trainMat, _tid);
-            }
-            assert booster != null;
-            _boosterHolder.update(booster);
-        } finally {
+    // This is called from driver
+    public Booster getBooster(BoosterParms boosterParms) {
+        final H2ONode boosterNode = getBoosterNode();
```

Comment:
Do you mean the getBoosterNode() function? That cost almost nothing (especially compared to actually extracting the Booster from XGBoost) - typically the first node of the cluster will have the data of the frame. If the cluster is larger then it is even more likely the first node will hold the data. For very small datasets it is possible that after some filtering the data will be located on the last node only. We can cache it but it will not have an effect IMHO